src/components/KymaCompanion/utils/formatMarkdown.tsx

```diff
@@ -1,48 +1,124 @@
 import { UI5Renderer } from 'components/KymaCompanion/components/Chat/messages/markedExtension';
-
 import Markdown from 'marked-react';
+import CodePanel from '../components/Chat/messages/CodePanel';
 
-export function formatMessage(text: string): JSX.Element[] {
-  const elements: JSX.Element[] = [];
-  let currentText: string = '';
-  let divs = 0;
-  let idx = 0;
-  let codeSection: boolean = false;
-  text.split('\n').forEach(line => {
-    if (line.trim().startsWith('<div')) {
-      divs += 1;
-      if (codeSection) {
-        return;
+export interface CodeSegmentLink {
+  address: string;
+  actionType: string;
+  name: string;
+}
+
+interface YamlBlock {
+  codeWithAction: true;
+  content: string;
+  link: CodeSegmentLink | null;
+  language: string;
+}
+
+export interface MarkdownText {
+  codeWithAction: false;
+  content: string;
+}
+
+type MessagePart = YamlBlock | MarkdownText;
+
+export function formatCodeSegment(text: string) {
+  const [language, ...lines] = text.split('\n');
+  return { language, code: lines.filter(line => line.trim()).join('\n') };
+}
+
+function extractYamlContent(block: string): string {
+  return block.match(/```([\s\S]*?)```/)?.[1]?.trim() || '';
+}
+
+function extractLink(block: string): CodeSegmentLink | null {
+  const match = block.match(
+    /<div class="link" link-type="(.*?)">\s*\[(.*?)\]\((.*?)\)\s*<\/div>/,
+  );
+  return match
+    ? {
+        name: match[2].trim(),
+        address: match[3].trim(),
+        actionType: match[1].trim(),
       }
-      elements.push(MarkdownText(currentText + '\n', idx));
-      currentText = line + '\n';
-      codeSection = true;
-      return;
-    }
+    : null;
+}
+
+function findClosingDiv(lines: string[], startIndex: number): number {
```

Comment:
Maybe
```suggestion
function findClosingDivIndex
```
would be a more fitting name, because the function returns an index